### ISSUE TITLE
`:mkspell` regression

### ISSUE THREAD

**Issue by julio-b** [neovim/neovim#23758]
### Problem

Latest versions of neovim cannot generate accurate .spl files for the greek language. I have tested many different dictionaries and they all seem to fail. Vim on the other hand can generate those spellfiles correctly.

### Steps to reproduce

Below is a Makefile that downloads the required dictionary files and executes `:mkspell` on them. Neovim will produce unusable .spl file without any word inside.
```Makefile
NVIM=nvim

all: el2.utf-8.spl
.PHONY: all clear clean

el2.utf-8.spl : el_GR.dic el_GR.aff
	$(NVIM) --headless --clean --cmd "mkspell! el2 el_GR" --cmd q

el.utf-8.spl : el_GR.dic el_GR.aff
	vim --clean --cmd "mkspell! el el_GR" --cmd q


el_GR.diff:
	curl -Lo el_GR.diff https://github.com/vim/vim/raw/master/runtime/spell/el/el_GR.diff
	$(NVIM) --headless --clean -c 'e ++enc=iso-8859-7 el_GR.diff' -c '3,47d' -c 'exe "4normal 62\<C-a>"' -c 'exe "5normal 62\<C-a>E62\<C-a>"' -c 'w ++enc=utf-8' -c 'q'

el_GR.aff: el_GR.diff
	curl -Lo el_GR.aff https://github.com/wooorm/dictionaries/raw/main/dictionaries/el/index.aff
	patch < el_GR.diff

el_GR.dic:
	curl -Lo el_GR.dic https://github.com/wooorm/dictionaries/raw/main/dictionaries/el/index.dic

clear:
	rm el_GR.{aff,dic,diff} el.utf-8.{spl,sug}

clean:
	rm el.utf-8.{spl,sug}

```

After `make` i get this output:
```
Reading affix file el_GR.aff...
Reading dictionary file el_GR.dic...
Compressing word tree...
Writing spell file el2.utf-8.spl...
Done!
Estimated runtime memory use: 0 bytes
Reading back spell file...
Performing soundfolding...
```
as you can see with `:spelldump` the file is completely empty.

`nvim --clean -c 'set spelllang=el2.utf-8.spl spell' -c 'spelldump'`

### Expected behavior

`make el.utf-8.spl` uses vim:
```
Reading affix file el_GR.aff...
Reading dictionary file el_GR.dic...
line      2, word      0 - ...line 660068, word 660066 - ...Compressing word tree...
Compressed case-folded: 6308160 of 6820007 nodes; 511847 (7%) remaining
Writing spell file el.utf-8.spl...
Done!
Estimated runtime memory use: 2559715 bytes
Reading back spell file...
Performing soundfolding...
Total number of words: 826887
Number of words after soundfolding: 755167
Compressing word tree...
Compressed case-folded: 4764952 of 5158331 nodes; 393379 (7%) remaining
Writing suggestion file el.utf-8.sug...
Estimated runtime memory use: 5065814 bytes
```

Both the .spl and .sug files are generated correctly, and `nvim --clean -c 'set spelllang=el.utf-8.spl spell' -c 'spelldump'` shows the full wordlist

### Neovim version (nvim -v)

NVIM v0.10.0-dev-386+ga6dd67f5b

### Vim (not Nvim) behaves the same?

no, vim 9.0.1572

### Operating system/version

archlinux

### Terminal name/version

kitty 0.28.1

### $TERM environment variable

xterm-kitty

### Installation

build from latest master

**Comment by zeertzjq** [comment]
In v0.9.0 this seems a segfault, but not in v0.10.0-dev.

**Comment by julio-b** [comment]
> In v0.9.0 this seems to segfault, but not in v0.10.0-dev.

I think the segfault was fixed in 7ddf235d7e
